Mathlib/Combinatorics/Enumerative/InclusionExclusion.lean

```diff
@@ -0,0 +1,111 @@
+/-
+Copyright (c) 2024 Yaël Dillies. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies
+-/
+import Mathlib.Algebra.BigOperators.Pi
+import Mathlib.Algebra.BigOperators.Ring
+import Mathlib.Algebra.Module.BigOperators
+import Mathlib.Algebra.Ring.NegOnePow
+
+/-!
+# Inclusion-exclusion principle
+
+This file proves several variants of the inclusion-exclusion principle.
+
+## TODO
+
+Use `Finset.prod_indicator_biUnion_sub_indicator` to prove the integral version of
+inclusion-exclusion.
+-/
+
+open Int
+
+namespace Finset
+variable {ι α G : Type*} [DecidableEq α] [AddCommGroup G] {s : Finset ι}
+
+lemma prod_indicator_biUnion_sub_indicator (hs : s.Nonempty) (f : ι → Finset α) (a : α) :
+    ∏ i ∈ s, (Set.indicator (s.biUnion f) 1 a - Set.indicator (f i) 1 a) = (0 : ℤ) := by
+  by_cases ha : a ∈ s.biUnion f
+  · obtain ⟨i, hi, ha⟩ := mem_biUnion.1 ha
+    exact prod_eq_zero hi <| by simp [*, -coe_biUnion]
+  · obtain ⟨i, hi⟩ := hs
+    have ha : a ∉ f i := fun h ↦ ha <| subset_biUnion_of_mem _ hi h
+    exact prod_eq_zero hi <| by simp [*, -coe_biUnion]
+
+/-- **Inclusion-exclusion principle** for the sum of a function over a union.
+
+The sum of a function `g` over the union of the `f i` over `i ∈ s` is the alternating sum of the
+sums of `g` over the intersections of the `f i`. -/
+theorem inclusion_exclusion_sum_biUnion (s : Finset ι) (f : ι → Finset α) (g : α → G) :
+    ∑ a ∈ s.biUnion f, g a = ∑ t : s.powerset.filter (·.Nonempty),
+      negOnePow (#t.1 + 1) • ∑ a ∈ t.1.inf' (mem_filter.1 t.2).2 f, g a := by
```

Comment:
Do we have a style guide about `negOnePow` vs `(-1 : Int) ^ _`? Here, you have a natural power and you don't need `Units Int`, so the latter would work.

Reply:
I thought I would need to specify that `-1` is an `Int`, but apparently not anymore 🎉